src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java

```diff
@@ -0,0 +1,344 @@
+            subResult = processTravelRequestsSubset(accessToken, testUserIdMappings, requestListing);
+            subResults.add(subResult);
+            page++;
+            currentQueryUrl = getQueryUrlForNextResultsPageIfPresent(requestListing);
+        } while (StringUtils.isNotBlank(currentQueryUrl));
+        
+        List<ConcurEventNotificationProcessingResultsDTO> results = subResults.build()
+                .flatMap(subResultStream -> subResultStream)
+                .collect(Collectors.toUnmodifiableList());
+        
+        LOG.info("processTravelRequests, Finished processing " + results.size() + " travel requests");
+        return results;
+    }
+
+    protected String getQueryUrlForNextResultsPageIfPresent(ConcurRequestV4ListingDTO requestListing) {
+        return Stream.of(requestListing.getOperations())
+                .filter(CollectionUtils::isNotEmpty)
+                .flatMap(List::stream)
+                .filter(operation -> StringUtils.equalsIgnoreCase(ConcurApiOperations.NEXT, operation.getName()))
+                .map(ConcurRequestV4OperationDTO::getHref)
+                .peek(this::verifyQueryUrlForNextResultsPageIsValid)
+                .findFirst()
+                .orElse(KFSConstants.EMPTY_STRING);
+    }
+
+    protected void verifyQueryUrlForNextResultsPageIsValid(String queryUrl) {
+        if (!ConcurUtils.validateFormatAndPrefixOfParameterizedUrl(queryUrl, getRequestV4Endpoint())) {
+            LOG.error("verifyQueryUrlForNextResultsPageIsValid, Detected an unexpected or malformed search URL: "
+                    + queryUrl);
+            throw new RuntimeException("Query URL for next page of search results is malformed");
+        }
+    }
+
+    protected ConcurRequestV4ListingDTO getPendingTravelRequests(String accessToken, int page, String queryUrl) {
+        String messageDetail = MessageFormat.format(
+                configurationService.getPropertyValueAsString(ConcurKeyConstants.MESSAGE_CONCUR_REQUESTV4_LISTING),
+                page);
+        return concurEventNotificationV2WebserviceService.buildConcurDTOFromEndpoint(
+                accessToken, queryUrl, ConcurRequestV4ListingDTO.class, messageDetail);
+    }
+
+    protected Stream<ConcurEventNotificationProcessingResultsDTO> processTravelRequestsSubset(
+            String accessToken, Map<String, String> testUserIdMappings, ConcurRequestV4ListingDTO requestListing) {
+        if (CollectionUtils.isEmpty(requestListing.getListItems())) {
+            return Stream.empty();
+        }
+        Stream.Builder<ConcurEventNotificationProcessingResultsDTO> subResults = Stream.builder();
+        
+        for (ConcurRequestV4ListItemDTO requestAsListItem : requestListing.getListItems()) {
+            if (StringUtils.isBlank(requestAsListItem.getId())) {
+                throw new IllegalStateException("Found a request item with a blank ID; this should NEVER happen");
```

Comment:
I agree this should never happen, but if it does happen, is there a way we can add some details here so we can identify the "bad" request.

Reply:
Ok, I'll update this to log some more detailed of the invalid request.